61_UI/include/nbl/this_example/renderpass/scene.hpp

```diff
@@ -0,0 +1,1362 @@
+	nbl::video::IGPUCommandBuffer::SClearDepthStencilValue depth = { .depth = 0.f };
+} clear;
+
+template<typename T, typename... Types>
+concept _implIsResourceTypeC = (std::same_as<T, Types> || ...);
+
+template<typename T, typename Types>
+concept RESOURCE_TYPE_CONCEPT = _implIsResourceTypeC<T, typename Types::DESCRIPTOR_SET_LAYOUT, typename Types::PIPELINE_LAYOUT, typename Types::RENDERPASS, typename Types::IMAGE_VIEW, typename Types::IMAGE, typename Types::BUFFER, typename Types::SHADER, typename Types::GRAPHICS_PIPELINE, typename Types::DESCRIPTOR_SET>;
+
+#define TYPES_IMPL_BOILERPLATE(WITH_CONVERTER) struct TYPES \
+{ \
+	using DESCRIPTOR_SET_LAYOUT = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUDescriptorSetLayout, nbl::video::IGPUDescriptorSetLayout>; \
+	using PIPELINE_LAYOUT = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUPipelineLayout, nbl::video::IGPUPipelineLayout>; \
+	using RENDERPASS = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPURenderpass, nbl::video::IGPURenderpass>; \
+	using IMAGE_VIEW = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUImageView, nbl::video::IGPUImageView>; \
+	using IMAGE = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUImage, nbl::video::IGPUImage>; \
+	using BUFFER = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUBuffer, nbl::video::IGPUBuffer>; \
+	using SHADER = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUShader, nbl::video::IGPUShader>; \
+	using GRAPHICS_PIPELINE = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUGraphicsPipeline, nbl::video::IGPUGraphicsPipeline>; \
+	using DESCRIPTOR_SET = std::conditional_t<WITH_CONVERTER, nbl::asset::ICPUDescriptorSet, nbl::video::IGPUDescriptorSet>; \
+}
+
+template<bool withAssetConverter>
+struct RESOURCES_BUNDLE_BASE
+{
+	TYPES_IMPL_BOILERPLATE(withAssetConverter);
+
+	struct REFERENCE_OBJECT
+	{
+		struct BINDINGS
+		{
+			nbl::asset::SBufferBinding<typename TYPES::BUFFER> vertex, index;
+		};
+
+		nbl::core::smart_refctd_ptr<typename TYPES::GRAPHICS_PIPELINE> pipeline = nullptr;
+
+		BINDINGS bindings;
+		nbl::asset::E_INDEX_TYPE indexType = nbl::asset::E_INDEX_TYPE::EIT_UNKNOWN;
+		uint32_t indexCount = {};
+	};
+
+	using REFERENCE_DRAW_HOOK = std::pair<REFERENCE_OBJECT, OBJECT_META>;
+
+	nbl::core::smart_refctd_ptr<typename TYPES::RENDERPASS> renderpass;
+	std::array<REFERENCE_DRAW_HOOK, EOT_COUNT> objects;
+	nbl::asset::SBufferBinding<typename TYPES::BUFFER> ubo;
+
+	struct
+	{
+		nbl::core::smart_refctd_ptr<typename TYPES::IMAGE_VIEW> color, depth;
+	} attachments;
+
+	nbl::core::smart_refctd_ptr<typename TYPES::DESCRIPTOR_SET> descriptorSet;
+};
+
+struct RESOURCES_BUNDLE : public RESOURCES_BUNDLE_BASE<false>
+{
+	using BASE_T = RESOURCES_BUNDLE_BASE<false>;
+};
+
+#define EXPOSE_NABLA_NAMESPACES() using namespace nbl; \
+using namespace core; \
+using namespace asset; \
+using namespace video; \
+using namespace scene; \
+using namespace system
```

Comment:
this is kinda horribly unreadable and complex, you need your `create` methods anyway

a simple `if constexpr` lower down would read much better

Reply:
bump

Reply:
resolved